### PR TITLE
extmod/modgzip.c: Add gzip.GzipFile providing compression and decompression.

### PR DESCRIPTION
**Pull request by jimmo**
Summary: Adds gzip/zlib/raw-deflate compression support. (See issues #7228, #5590, and earlier PRs #8195, #5613).

Historically MicroPython provided a CPython-compatible implementation of `zlib.decompress` as well as a MicroPython-specific `zlib.DecompIO` which is similar to CPython's `zlib.decompressobj` except more useful as a stream wrapper, for example to decompress a file or socket, because you pull (read) decompressed data out of it, rather than pushing (write) compressed data into it.

However, CPython also provides `gzip.GzipFile` which works exactly like MicroPython's `DecompIO`, the only limitation is that it only supports gzip-format data (whereas `DecompIO` supported the same `wbits` argument that `zlib.decompress` supported).

This PR removes the `zlib` module, and replaces it with a `gzip` module providing a CPython-compatible `gzip.GzipFile` (now supporting decompression _and_ compression), but also with a MicroPython-extension allowing `GzipFile` to read/write zlib and raw-deflate files (via an optional, MicroPython-specific `wbits` keyword argument).

For backwards compatibility, I have written a pure-Python implementation of the functionality of the former `zlib` module (including now `zlib.compress`), which will be published to micropython-lib (https://github.com/micropython/micropython-lib/pull/694) and can be optionally installed via mip (or frozen). It's only about 16 lines (351 byte .mpy file) but adds +428 bytes when frozen. The argument for not including it by default is:
- (De)compression isn't a feature that everyone uses.
- Even in applications that use (de)compression, GzipFile provides a much more memory-efficient way to work with reading/writing compressed data to/from a stream (which is the typical use case).

This PR includes the following changes:
- Add a shared `mp_stream___exit___obj` that avoids each file/stream-like object needing to implement this themselves. (This is a code size optimisation saving ~200 bytes)
- Add a memory-efficient implementation of LZ77 compression to lib/uzlib (Written by @dpgeorge ).
- General code size optimisations for lib/uzlib. (Saves around 650 bytes)
- Code cleanup and simplication for lib/uzlib.
- Remove the `zlib` module.
- Add the `gzip` module.
- Add tests for `gzip.GzipFile`.
- Update docs for `zlib` to describe the micropython-lib version which wraps `GzipFile`.
- Add docs for `gzip`.

Overall this PR adds +1076 bytes to PYBV11 (i.e. adding compression support). This is made up of -3012 (remove zlib), +2900 (add decompression-only GzipFile), +1188 (add compression to GzipFile).

Compared to #8195 & #5613 which are quite similar (cc @harbaum, @andrewleech) this PR adds support for streaming as well as non-gzip (i.e. zlib/raw) formats and a code size reduction as well as reduced memory usage in the compressor.

_This work was funded through GitHub Sponsors, by specific request from a supporter._

### CODE REVIEW

**Comment by jimmo** [comment]
The main TODO is whether to enable compression by default. Currently it's set to the "extra features" level which is the same as decompression support. I would be quite happy to move it to "full features".

File: docs/library/gzip.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+**Availability:**
+* Added in MicroPython v1.21.
+* Decompression: Enabled via the ``MICROPY_PY_GZIP`` build option, on by default on ports with the "extra features" level or higher (most boards).
+* Compression: Enabled via the ``MICROPY_PY_GZIP_COMPRESS`` the build option. **TODO**
+
+Functions
+---------
+
+.. class:: GzipFile(fileobj, mode, wbits=25, hist=None, /)
```
Comment (dpgeorge) [inline docs/library/gzip.rst:17]: I think this should be `GzipFile(*, fileobj, mode, wbits=25, hist=None)`

File: extmod/modgzip.c [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+    // Unsupported from CPython: filename, compresslevel, mtime=None
+    // MicroPython extras: wbits, hist
+
+    // Parse args
+    enum { ARG_fileobj, ARG_mode, ARG_wbits, ARG_hist };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_fileobj, MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_mode, MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_wbits, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 25} }, // Default to gzip with smallest window.
+        { MP_QSTR_hist, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
Comment (dpgeorge) [inline extmod/modgzip.c:209]: The default should be `.u_rom_obj = MP_ROM_NONE` to match the docs signature.

**Comment by robert-hh** [comment]
There is not much difference ATM between the EXTRA and FULL feature setting. It seems to be only detailed error reporting, and either setting produces the same code size for a SAMD51 and 432 byte difference for a Teensy 4.0. MIMXRT and SAMD51 are the only boards with FULL setting, whether needed or not.
Adding GZIP compression increases the code size for the boards by about 1k. That seems to be acceptable for boards which support EXTRA features, without having checked the code space exhausting for every possible port & board. Including GZIP in the FULL feature level set would at least avoid a possible code size conflict for the moment, but would require to adding at a per-port or per-board level is required.

**Comment by dpgeorge** [comment]
It's worth noting that the compression implemented here uses much less memory compared to #8195 and #5613.  In particular it doesn't allocate a large hash table to store indices into the history of the previously seen text.  Instead it simply does a brute-force-search of the history text to find a match for the compressor.  Yes this is slower (linear search vs hash table lookup) but with a small enough history (eg 512 bytes) it's not that slow.  And a small history does not impact the compression too much.

To give some more concrete numbers comparing memory use between the approaches:
- standard approach (eg #8195): inplace compression, all text to compress must be in RAM (or at least memory addressable), and then an additional 16k bytes RAM of hash table pointers, pointing into the text
- the approach in this PR: streaming compression, only a limited amount of previous text must be in RAM (user selectable, defaults to 512 bytes)

So if you want to compress, say, 1k of data, the standard approach requires all that data to be in RAM, plus an additional 16k of RAM for the hash table pointers.  In this PR, you only need the 1k of data in RAM.  Or if it's streaming from a file (or elsewhere), you could get away with only 256 bytes of RAM for the sliding history and still get very decent compression.

In summary: because compression takes such a large amount of RAM (in the traditional case) and it's not really suitable for microcontrollers, the approach taken in this PR is to minimise RAM usage as much as possible, and still have acceptable performance (speed and compression ratio).

File: extmod/modgzip.c [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+            // trailing checksum.
+            *wbits = -*wbits;
+        } else if (*wbits >= 9 && *wbits <= 15) {
+            // +9 to +15: The base-two logarithm of the window size, which
+            // therefore ranges between 512 and 32768. Larger values produce
+            // better compression at the expense of greater memory usage. The
+            // resulting output will include a zlib-specific header and trailer.
+            self->mode = GZIPFILE_MODE_WRITE_ZLIB;
+            self->state[0].write.input_checksum = 1; // ADLER32
+        } else if (*wbits >= 25 && *wbits <= 31) {
```
Comment (dpgeorge) [inline extmod/modgzip.c:173]: I reckon it's worth allowing a smaller minimum wbits, to get a smaller history.  Eg 22?  Then the 9's about could be 6's.

File: lib/uzlib/lz77.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    return longest_len;
+}
+
+// Compress the given chunk of data.
+void uzlib_lz77_compress(uzlib_lz77_state_t *state, const uint8_t *src, unsigned len) {
+    bool use_src_as_history = false;
+    if (state->hist_buf == NULL) {
```
Comment (dpgeorge) [inline lib/uzlib/lz77.c:62]: This case is the in-place scenario, where the history is exactly the input text.

File: extmod/modgzip.c (hunk 3 above)
Comment (jimmo) [reply extmod/modgzip.c]: I think this is worthwhile as it's a significant memory saving for small inputs, even if it is another CPython diff.

I've implemented this in a separate commit, but it can be squashed later.

File: lib/uzlib/lz77.c (hunk 4 above)
Comment (jimmo) [reply lib/uzlib/lz77.c]: Updated. Removing this saves 64 bytes.

It was a neat feature for `zlib.compress` case, but I really think we should be encouraging use of the streaming interface everywhere (and I've tried to convey that in the docs).

**Comment by jimmo** [comment]
Here is a test to show that the changes to defl_static.c produce the exact same output (by calling `uzlib_literal` and `uzlib_match` with all possible inputs).

https://gist.github.com/jimmo/90b5bbc4ffc813890bc4e9231e22338f

**Comment by jimmo** [comment]
Added more tests based on the coverage results.

**Comment by jimmo** [comment]
Replaced by https://github.com/micropython/micropython/pull/11905

**Comment by dpgeorge** [comment]
Closing in favour of #11905.
